<!DOCTYPE html>
<html lang="en">

<head>

  <title>Evaluating 515F primers</title>
  <meta itemprop="description"
        name="description"
        content="Before we begin, I stumbled across this really great blog post from the Knight lab …"/>

  
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>

  <!-- Open Graph tags -->
  <meta property="og:locale" content="en_US">
  <meta property="og:site_name" content="A microbiologist with a data science problem"/>
  <meta property="article:author" content="Chuck Pepe-Ranney"/>

  <!-- Twitter card tags -->
  <meta name="twitter:card" content="summary"/>

  <meta name="twitter:creator" content="@chuckpr"/>


  <link rel="stylesheet" href="https://unpkg.com/tachyons@4.10.0/css/tachyons.min.css"/>

  <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/theme/css/site.css">

  <link rel="shortcut icon" type="image/x-icon" href="https://chuckpr.github.io/favicon.ico"/>
  <link rel="icon" type="image/x-icon" href="https://chuckpr.github.io/favicon.ico"/>



  <!-- Open Graph tags -->
  <meta property="og:type" content="article">
  <meta property="og:title" content="Evaluating 515F primers"/>
  <meta property="og:url" content="https://chuckpr.github.io/posts/primers2/"/>
  <meta property="og:image" content="https://res.cloudinary.com/chuckpr-github-io/image/upload/co_rgb:f2f2f2,w_1000,c_fit,g_east,x_100,l_text:Helvetica_90_right:Evaluating%20515F%20primers/blog_card_jieod8.png"/>
  <meta property="article:published_time" content="2015-04-17 00:00:00-04:00"/>
  <meta property="og:description"
        content="Before we begin, I stumbled across this really great blog post from the Knight lab …"/>

  <!-- Twitter card tags -->
  <meta name="twitter:title" content="Evaluating 515F primers">
  <meta name="twitter:description"
        content="Before we begin, I stumbled across this really great blog post from the Knight lab …"/>

  <link rel="canonical" href="https://chuckpr.github.io/posts/primers2/">

  <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/theme/css/pygments.css">
  <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/theme/css/article.css">
    <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/extra/css/notebook.css">

</head>

<body>
  <header class="flex flex-column avenir mw7 ml-auto mr-auto">
<div class="pt5 pb2 pl0-l pl2">
  <a href="https://chuckpr.github.io/" class="link dim f2-ns f3 fw4 blue tracked-tight lh-title">
    A microbiologist with a data science problem
  </a>
</div>  </header>
  <main class="mw7 ml-auto mr-auto ph3 flex flex-column min-vh-100">
  <article class="flex flex-column">
    <section id="article-title" class="flex justify-between mt4 avenir bb b--moon-gray pb2">
      <div>
        <div class="flex flex-column">
          <div>
            <p class="f3-ns f4 mv0 lh-title">
              Evaluating 515F primers
            </p>
          </div>
          <div class="ml3">
            <p class="f6 ttc mb0 mt1 mid-gray">
              Fri 17 April 2015
            </p>
          </div>
        </div>
      </div>
      <div class="w-40">
        <div class="flex flex-wrap justify-end">
<div class="mv1 mr2 bg-light-green br3 pa2 grow">
  <a href="https://chuckpr.github.io/tag/r" class="link dark-pink underline-hover fw5">
      #R
  </a>
</div><div class="mv1 mr2 bg-light-green br3 pa2 grow">
  <a href="https://chuckpr.github.io/tag/python" class="link dark-pink underline-hover fw5">
      #Python
  </a>
</div><div class="mv1 mr2 bg-light-green br3 pa2 grow">
  <a href="https://chuckpr.github.io/tag/amplicon" class="link dark-pink underline-hover fw5">
      #amplicon
  </a>
</div>        </div>
      </div>
    </section>
    <section id="article-content" class="ph3 garamond near-black" style="font-size: 17px">
      <p>Before we begin, I stumbled across <a href="http://microbe.net/2014/10/22/comparing-the-new-16s-rrna-v4-and-its-primers-to-the-old-primers-results/"><strong>this really great blog post</strong></a> from the Knight lab that you should <strong>definitely</strong> check out if you're interested in SSU rRNA gene primers. They have some experimental data up which is great to see.</p>



<p>Let's start by summarizing a few key findings from the <a href="https://chuckpr.github.io/posts/primers1/">last post</a>.</p>



<ol>
<li>The 515F target in Bacteria is extremely conserved -- one sequence is found many many times more than any other.</li>
<li>Although the bacterial 515F targets are predominantly one sequence, the second and third most abundant bacterial 515F targets differ from the most abundant target near the 3' end.</li>
<li>The 515F targets in Archaea and Bacteria have little overlap.</li>
<li>Archaeal 515F targets are more evenly distributed than bacterial 515F targets. </li>
<li>The Nanoarchaea appear to have a diagnostic 515F sequence. </li>
</ol>



<p>Based on these findings we might ask a few questions:</p>



<ul>
<li>Does the 515F primer account for the 3' differences between the three most abundant bacterial targets? If not, what are the implications?</li>
<li>Does 515F hit Archaea <strong>and</strong> Bacteria equally well?</li>
<li>Does 515F hit Nanoarchaea in particular?</li>
</ul>



<p>The questions above imply one version of 515F. In reality, the 515F primer has been modified over time so we should address those questions over a representative collection of 515F primers.</p>



<p>In this post we'll look at two 515F versions and propose an updated version and see how it compares. Let's evaluate the 515F version used by <a href="http://www.nature.com/ismej/journal/v5/n5/full/ismej2010171a.html">Bates <em>et al.</em> (2010)</a> because the authors provide a nice primer analysis in their paper. We'll also look at the 515F version that is on the <a href="http://www.earthmicrobiome.org/emp-standard-protocols/16s/">Earth Microbiome Project (EMP)</a> website.</p>



<p>I'm just going to use the <code>regex</code> <code>Python</code> module (<a href="https://pypi.python.org/pypi/regex">here</a>) to evaluate the primers. We need to represent the primers with the degeneracy in regular expression syntax. The <code>{e}</code> at the end will allow us to enable fuzzy matching.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">versions</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;EMP&quot;</span> <span class="p">:</span> <span class="s2">&quot;(GTG[CT]CAGC[CA]GCCGCGGTAA)</span><span class="si">{e}</span><span class="s2">&quot;</span><span class="p">,</span> 
            <span class="s2">&quot;Bates&quot;</span> <span class="p">:</span> <span class="s2">&quot;(GTGCCAGC[CA]GCCGCGGTAA)</span><span class="si">{e}</span><span class="s2">&quot;</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Ok, let's bring in the <code>Python</code> modules we need and some IPython extensions.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span>
<span class="o">%</span><span class="n">load_ext</span> <span class="n">rpy2</span><span class="o">.</span><span class="n">ipython</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">align</span> <span class="kn">import</span> <span class="n">aligner</span>
<span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">parse_fasta</span><span class="p">,</span> <span class="n">SequenceCollection</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">regex</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">);</span> <span class="nf">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">magrittr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_text output_subarea ">
<pre>Loading required package: grid
</pre>
</div>

</div>

</div>
</div>

</div>

<p>We'll make the same 515F target dataframe as in the <a href="https://chuckpr.github.io/posts/primers1/">previous post</a>.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">parse_tax</span><span class="p">(</span><span class="n">line</span><span class="p">):</span>
    <span class="nb">id</span><span class="p">,</span> <span class="n">tax_str</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">k</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">o</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">s</span> <span class="o">=</span> <span class="n">tax_str</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
    <span class="n">tax_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;kingdom&quot;</span> <span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;phylum&quot;</span>  <span class="p">:</span> <span class="n">p</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;class&quot;</span>   <span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;family&quot;</span>  <span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;order&quot;</span>   <span class="p">:</span> <span class="n">o</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;genus&quot;</span>   <span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;species&quot;</span> <span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;__&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;id&quot;</span>      <span class="p">:</span> <span class="nb">id</span><span class="p">}</span>
    <span class="k">return</span> <span class="n">tax_dict</span>
    

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">([</span><span class="n">parse_tax</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> 
                                <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data/gg_13_5_otus/taxonomy/99_otu_taxonomy.txt&quot;</span><span class="p">)])</span>

<span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/gg_13_5_otus/rep_set/99_otus.fasta&quot;</span>
<span class="n">degens</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">iupac_degenerate_characters</span><span class="p">()</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">SequenceCollection</span><span class="o">.</span><span class="n">from_fasta_records</span><span class="p">([(</span><span class="n">n</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span> 
                                              <span class="k">if</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="n">s</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">degens</span><span class="p">)],</span> 
                                             <span class="n">DNA</span><span class="p">)</span>

<span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()]</span>

<span class="n">F515</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;GTGCCAGCCGCCGCGGTAA&quot;</span><span class="p">)</span> <span class="c1">#GTGCCAGCMGCCGCGGTAA  M = [CA]</span>

<span class="n">primer_targets_515</span> <span class="o">=</span> <span class="p">[</span><span class="n">aligner</span><span class="p">(</span><span class="n">F515</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> 
                              <span class="n">seq</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
                              <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;glocal&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> 
                      <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()]</span>

<span class="n">df_targets</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">primer_targets_515</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">ids</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;515F&quot;</span><span class="p">),</span> <span class="n">how</span> <span class="o">=</span> <span class="s2">&quot;left&quot;</span><span class="p">)</span>
<span class="n">df_targets</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;515F&quot;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># &lt;!-- collapse=True --&gt;</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Ok, now the fun stuff. Let's define a function that counts substitutions between the target and the primer. If there's an indel in the target, this function returns an empty string. Eventually we'll pass this dataframe into R and it's easy to replace empty strings with R NA values. One quick note: we'll just be working with the ten most abundant targets in Bacteria and Archaea for the following figures.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">count_subs</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">primer_regex</span><span class="p">):</span>
    <span class="n">subs</span><span class="p">,</span> <span class="n">ins</span><span class="p">,</span> <span class="n">dels</span> <span class="o">=</span> <span class="n">regex</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">primer_regex</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span><span class="o">.</span><span class="n">fuzzy_counts</span>
    <span class="k">if</span> <span class="n">ins</span> <span class="ow">or</span> <span class="n">dels</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">subs</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Now we'll add a column to the target dataframe for each primer version. The values are the number of substitutions between the target and the primer.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">versions</span><span class="p">:</span>
    <span class="n">df_targets</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_targets</span><span class="p">[</span><span class="s2">&quot;515F&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">count_subs</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">versions</span><span class="p">[</span><span class="n">key</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Push the dataframe into R and we're ready to go! I want to make a figure that conveys three bits of information:</p>



<ol>
<li>The target,</li>
<li>the number of mismatches between the primer and the target, and</li>
<li>the representation of the target in selected taxonomic groups.</li>
</ol>



<p>So, let's make a heatmap (sort of) figure where columns are targets (10 most abundant archaeal and bacterial), rows represent 515F primer versions and the color of each cell represents the mismatches between primer and target. We can scale the size of each cell to the count of each target within the given taxon (log scaled).</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%</span><span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">df_targets</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">w</span> <span class="m">650</span>
<span class="n">d</span> <span class="o">=</span> <span class="n">df_targets</span> <span class="o">%&gt;%</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">kingdom</span><span class="p">,</span> <span class="n">X515F</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">summarize</span><span class="p">(</span><span class="n">count</span> <span class="o">=</span> <span class="nf">n</span><span class="p">(),</span> <span class="n">EMP</span> <span class="o">=</span> <span class="nf">first</span><span class="p">(</span><span class="n">EMP</span><span class="p">),</span> <span class="n">Bates</span> <span class="o">=</span> <span class="nf">first</span><span class="p">(</span><span class="n">Bates</span><span class="p">))</span> <span class="o">%&gt;%</span>
    <span class="nf">filter</span><span class="p">(</span><span class="nf">rank</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">count</span><span class="p">),</span> <span class="n">ties.method</span> <span class="o">=</span> <span class="s">&quot;random&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="m">10</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">arrange</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">count</span><span class="p">),</span> <span class="n">kingdom</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">X515F</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">X515F</span><span class="p">,</span> <span class="n">levels</span> <span class="o">=</span> <span class="nf">unique</span><span class="p">(</span><span class="n">.</span><span class="o">$</span><span class="n">X515F</span><span class="p">)))</span> <span class="o">%&gt;%</span> 
    <span class="p">{</span>
        <span class="n">.</span><span class="p">[</span><span class="n">.</span><span class="o">==</span><span class="s">&quot;&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">NA</span>
        <span class="n">.</span>
    <span class="p">}</span> <span class="o">%&gt;%</span>
    <span class="nf">gather</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="o">-</span><span class="n">count</span><span class="p">,</span> <span class="o">-</span><span class="n">kingdom</span><span class="p">,</span> <span class="o">-</span><span class="n">X515F</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">X515F</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">key</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">count</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">value</span><span class="p">))</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="n">kingdom</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">scales</span> <span class="o">=</span> <span class="s">&quot;free_y&quot;</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">(</span><span class="n">shape</span> <span class="o">=</span> <span class="m">22</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">scale_size_area</span><span class="p">(</span><span class="n">max_size</span> <span class="o">=</span> <span class="m">14</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="nf">brewer.pal</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="s">&quot;RdYlBu&quot;</span><span class="p">),</span> <span class="n">na.value</span> <span class="o">=</span> <span class="s">&quot;grey70&quot;</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;# mismatches&quot;</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">theme_bw</span><span class="p">()</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span> <span class="o">=</span> <span class="m">65</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">14</span><span class="p">,</span> <span class="n">family</span> <span class="o">=</span> <span class="s">&quot;mono&quot;</span><span class="p">),</span>
              <span class="n">axis.text.y</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">12</span><span class="p">),</span>
              <span class="n">axis.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">),</span>
              <span class="n">strip.background</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
              <span class="n">strip.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">),</span>
              <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">12</span><span class="p">),</span>
              <span class="n">legend.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">))</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">guides</span><span class="p">(</span><span class="n">fill</span> <span class="o">=</span> <span class="nf">guide_legend</span><span class="p">(</span><span class="n">override.aes</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)),</span> <span class="n">size</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="n">p</span>

<span class="c1"># &lt;!-- collapse=True --&gt;</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/primers2_22_0.png"
>
</div>

</div>

</div>
</div>

</div>

<p>Cool! You can see that the EMP primer has improved zero mismatch coverage of Archaea (also discussed in <a href="http://microbe.net/2014/10/22/comparing-the-new-16s-rrna-v4-and-its-primers-to-the-old-primers-results/">this post</a>). There are still mismatches at the 3' end between second and third most abundant bacterial 515F targets with each primer version.</p>



<p>Let's explore what the implications of the 3' end mismatch between primers and bacterial targets might be. I want to look at how each primer hits bacterial phyla. Let's define a function to make this plot. The function will take a couple arguments. One argument will specify the taxon to plot (right now just Bacteria or Archaea) and the other will specify the primer. In this figure rows will be targets (10 most abundant targets) and columns will be taxa. The colors will represent mismatches against the selected primer. Size will be based on relative abundance for the target within the given column. <strong>These figures are somewhat small but you can zoom in by right clicking and opening the figure in a new browser window or tab.</strong></p>


<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">w</span> <span class="m">1300</span> <span class="o">-</span><span class="n">h</span> <span class="m">1000</span>
<span class="n">primer_plot</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">taxon</span> <span class="o">=</span> <span class="s">&quot;Bacteria&quot;</span><span class="p">,</span> <span class="n">primer</span> <span class="o">=</span> <span class="s">&quot;EMP&quot;</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">d</span> <span class="o">=</span> <span class="n">df_targets</span> <span class="o">%&gt;%</span>
        <span class="p">{</span>
            <span class="n">.</span><span class="p">[</span><span class="n">.</span><span class="o">==</span><span class="s">&quot;&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">NA</span>
            <span class="n">.</span>
        <span class="p">}</span> <span class="o">%&gt;%</span>
        <span class="nf">filter</span><span class="p">(</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">phylum</span><span class="p">),</span> <span class="n">kingdom</span> <span class="o">==</span> <span class="n">taxon</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="p">{</span>
            <span class="n">psort</span> <span class="o">=</span> <span class="nf">group_by</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="n">phylum</span><span class="p">)</span> <span class="o">%&gt;%</span> 
                <span class="nf">summarize</span><span class="p">(</span><span class="n">count</span> <span class="o">=</span> <span class="nf">n</span><span class="p">())</span> <span class="o">%&gt;%</span> 
                <span class="nf">arrange</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">count</span><span class="p">))</span> <span class="o">%&gt;%</span>
                <span class="nf">extract2</span><span class="p">(</span><span class="s">&quot;phylum&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">as.character</span>
            <span class="nf">mutate</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="n">phylum</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">phylum</span><span class="p">,</span> <span class="n">levels</span> <span class="o">=</span> <span class="n">psort</span><span class="p">))</span>
        <span class="p">}</span> <span class="o">%&gt;%</span>
        <span class="nf">group_by</span><span class="p">(</span><span class="n">phylum</span><span class="p">,</span> <span class="n">X515F</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="nf">do</span><span class="p">(</span><span class="nf">summarize</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="n">count</span> <span class="o">=</span> <span class="nf">n</span><span class="p">(),</span> <span class="n">primer</span> <span class="o">=</span> <span class="nf">first</span><span class="p">(</span><span class="n">.</span><span class="p">[[</span><span class="n">primer</span><span class="p">]])))</span> <span class="o">%&gt;%</span>
        <span class="nf">group_by</span><span class="p">(</span><span class="n">phylum</span><span class="p">)</span> <span class="o">%&gt;%</span>
        <span class="nf">mutate</span><span class="p">(</span><span class="n">count.relative</span> <span class="o">=</span> <span class="n">count</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">count</span><span class="p">))</span> <span class="o">%&gt;%</span>
        <span class="p">{</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="nf">group_by</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="n">X515F</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="nf">summarize</span><span class="p">(</span><span class="n">tmp</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="n">count</span><span class="p">))</span> <span class="o">%&gt;%</span>
                <span class="nf">filter</span><span class="p">(</span><span class="nf">rank</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">tmp</span><span class="p">),</span> <span class="n">ties.method</span> <span class="o">=</span> <span class="s">&quot;random&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="m">10</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="nf">arrange</span><span class="p">(</span><span class="n">tmp</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="nf">extract2</span><span class="p">(</span><span class="s">&quot;X515F&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="n">as.character</span>
            <span class="nf">filter</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="n">X515F</span> <span class="o">%in%</span> <span class="n">keep</span><span class="p">)</span> <span class="o">%&gt;%</span>
                <span class="nf">mutate</span><span class="p">(</span><span class="n">X515F</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">X515F</span><span class="p">,</span> <span class="n">levels</span> <span class="o">=</span> <span class="n">keep</span><span class="p">),</span> <span class="n">primer</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">primer</span><span class="p">,</span> <span class="n">levels</span> <span class="o">=</span> <span class="m">0</span><span class="o">:</span><span class="m">4</span><span class="p">))</span>
        <span class="p">}</span>


    <span class="n">p</span> <span class="o">=</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">X515F</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">phylum</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="n">count.relative</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">primer</span><span class="p">))</span>
               
    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">(</span><span class="n">shape</span> <span class="o">=</span> <span class="m">22</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="m">0.75</span><span class="p">)</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;# mismatches&quot;</span><span class="p">)</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="nf">brewer.pal</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="s">&quot;RdYlBu&quot;</span><span class="p">),</span> <span class="n">na.value</span> <span class="o">=</span> <span class="s">&quot;grey70&quot;</span><span class="p">)</span>
    
    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">ggtitle</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Figure for the &quot;</span><span class="p">,</span> <span class="n">primer</span><span class="p">,</span> <span class="s">&quot;515F primer&quot;</span><span class="p">))</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">theme_bw</span><span class="p">()</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span> <span class="o">=</span> <span class="m">-90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">vjust</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">12</span><span class="p">),</span>
                  <span class="n">axis.text.y</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">,</span> <span class="n">family</span> <span class="o">=</span> <span class="s">&quot;mono&quot;</span><span class="p">),</span>
                  <span class="n">axis.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">),</span>
                  <span class="n">strip.background</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
                  <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">12</span><span class="p">),</span>
                  <span class="n">legend.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">),</span>
                  <span class="n">strip.text</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">())</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">guides</span><span class="p">(</span><span class="n">fill</span> <span class="o">=</span> <span class="nf">guide_legend</span><span class="p">(</span><span class="n">override.aes</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">5</span><span class="p">)),</span> <span class="n">size</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="nf">coord_flip</span><span class="p">()</span>

    <span class="n">p</span>
<span class="p">}</span>

<span class="nf">grid.arrange</span><span class="p">(</span><span class="nf">primer_plot</span><span class="p">(),</span> <span class="nf">primer_plot</span><span class="p">(</span><span class="n">primer</span> <span class="o">=</span> <span class="s">&quot;Bates&quot;</span><span class="p">),</span> <span class="n">nrow</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>

<span class="c1"># &lt;!-- collapse=True --&gt;</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/primers2_25_0.png"
>
</div>

</div>

</div>
</div>

</div>

<p>As you can see, the 3' mismatch will probably effect recovery of <strong>TM7</strong> as well as a smattering of sequences throughout the most abundant phyla.</p>



<p>We can make the same plot for Archaea. This highlights how the EMP 515F version hits the Crenarchaea with zero mismatches but neither primer matches the Nanoarchaea targets well.</p>


<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">w</span> <span class="m">1200</span> <span class="o">-</span><span class="n">h</span> <span class="m">400</span>
<span class="nf">grid.arrange</span><span class="p">(</span><span class="nf">primer_plot</span><span class="p">(</span><span class="n">taxon</span> <span class="o">=</span> <span class="s">&quot;Archaea&quot;</span><span class="p">),</span> 
             <span class="nf">primer_plot</span><span class="p">(</span><span class="n">taxon</span> <span class="o">=</span> <span class="s">&quot;Archaea&quot;</span><span class="p">,</span> <span class="n">primer</span> <span class="o">=</span> <span class="s">&quot;Bates&quot;</span><span class="p">),</span> <span class="n">ncol</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/primers2_28_0.png"
>
</div>

</div>

</div>
</div>

</div>

<p>It's a bit unrealistic to design the perfect primer. But, it's certainly possible to quantify and explore <em>in silico</em> primer bias and to interpret results with the bias in mind. It's really great to see how the EMP folks have improved the 515F primer to hit more Crenarchaea. One suggestion from our analysis today is to add degeneracy at the 3' end to account for TM7 and many members of more abundant groups. The current EMP 515F primer is <strong>5'-GTGYCAGCMGCCGCGGTAA-3'</strong> but it might be useful to change that to <strong>5'-GTGYCAGCMGCCGCGGTVA-3'</strong>.</p>



<p>I'm also wondering if you could just chop off the last two nucleotides? This would drop the annealing temp but perhaps it would be negligible(?).</p>



<p><strong>Here is what the proposed 515F looks like against Bacteria.</strong> You can see that it hits TM7 and the second most abundant target with zero mismatches.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df_targets</span><span class="p">[</span><span class="s2">&quot;proposed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_targets</span><span class="p">[</span><span class="s2">&quot;515F&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">count_subs</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;(GTG[CT]CAGC[CA]GCCGCGGT[AGC]A)</span><span class="si">{e}</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="o">%</span><span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">df_targets</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">w</span> <span class="m">1300</span> <span class="o">-</span><span class="n">h</span> <span class="m">500</span>
<span class="nf">primer_plot</span><span class="p">(</span><span class="n">primer</span> <span class="o">=</span> <span class="s">&quot;proposed&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/primers2_33_0.png"
>
</div>

</div>

</div>
</div>

</div>

<p>And we can address the questions from above.</p>



<ul>
<li>Do 515F primers account for the 3' differences between the three most abundant targets? If not, what are the implications?</li>
</ul>



<blockquote><p>No. Current 515F primers are biased against TM7 and have mismatches at the 3' end with sequences throughout the most commonly found bacterial phyla (although current primers do hit the most abundant target in Archaea and Bacteria with zero mismatches and there's only one mismatch between current targets and most bacterial targets). So, on the whole, the latest primers are great but still room for improvement.</p>
</blockquote>



<ul>
<li>Does 515F hit Archaea <strong>and</strong> Bacteria equally well?</li>
</ul>



<blockquote><p>I would say yes, tentatively. The latest EMP primer appears to hit most Archaea with zero mismatches. It also hits most Bacteria with one mismatch or less.</p>
</blockquote>



<ul>
<li>Does 515F hit Nanoarchaea in particular?</li>
</ul>



<blockquote><p>No. Nanoarchaea have distinct 515F targets and 515F primers are biased against Nanoarchaea <em>in silico.</em></p>
</blockquote>


 


    </section>
<div id="disqus_thread" class="mt4">
</div>
<script>

  var disqus_config = function () {
    this.page.url = 'https://chuckpr.github.io/posts/primers2/';
    this.page.identifier = 'posts/primers2/';
  };

  (function() { // DON'T EDIT BELOW THIS LINE
    var d = document, s = d.createElement('script');
    s.src = 'https://chuckpr-blog.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();

</script>
<noscript>
  Please enable JavaScript to view the
  <a href="https://disqus.com/?ref_noscript">
    comments powered by Disqus.
  </a>
</noscript>  </article>
  </main>
  <footer class="pv3 flex justify-between mt4 mb2 mw7 ml-auto mr-auto avenir">
    <div class="flex flex-column pl3">
      <div class="mid-gray fw4 f6-ns f7 mb2">
        Copyright &copy; 2020 Chuck&nbsp;Pepe&#8209;Ranney
      </div>
      <div class="mid-gray" style="font-size: 0.65rem;">
        Powered&nbsp;by&nbsp;<a href="https://getpelican.com/" class="link dim
          mid-gray f6 fw6">Pelican</a>,
        Theme&nbsp;by&nbsp;<a href="https://github.com/chuckpr/simple-pelican-blog-theme"
          class="link dim mid-gray fw6">@chuckpr</a>
      </div>
    </div>
    <div class="flex justify-end w-50-ns w-30 mr4-ns mr0 ml2">
            <div class="w-10-ns w2 ml2 grow">
              <a href="https://github.com/chuckpr" class="light-purple dim link">
                <div class="o-50 pa1-l">
<svg viewBox="0 0 24 24" fill="currentColor">
  <title>GitHub icon</title>
  <path d="M12 .297c-6.63 0-12 5.373-12 12 0 5.303 3.438 9.8 8.205 11.385.6.113.82-.258.82-.577 0-.285-.01-1.04-.015-2.04-3.338.724-4.042-1.61-4.042-1.61C4.422 18.07 3.633 17.7 3.633 17.7c-1.087-.744.084-.729.084-.729 1.205.084 1.838 1.236 1.838 1.236 1.07 1.835 2.809 1.305 3.495.998.108-.776.417-1.305.76-1.605-2.665-.3-5.466-1.332-5.466-5.93 0-1.31.465-2.38 1.235-3.22-.135-.303-.54-1.523.105-3.176 0 0 1.005-.322 3.3 1.23.96-.267 1.98-.399 3-.405 1.02.006 2.04.138 3 .405 2.28-1.552 3.285-1.23 3.285-1.23.645 1.653.24 2.873.12 3.176.765.84 1.23 1.91 1.23 3.22 0 4.61-2.805 5.625-5.475 5.92.42.36.81 1.096.81 2.22 0 1.606-.015 2.896-.015 3.286 0 .315.21.69.825.57C20.565 22.092 24 17.592 24 12.297c0-6.627-5.373-12-12-12"/>
</svg>                </div>
              </a>
            </div>
            <div class="w-10-ns w2 ml2 grow">
              <a href="https://www.linkedin.com/in/chuckpr/" class="light-purple dim link">
                <div class="o-50 pa1-l">
<svg viewBox="0 0 24 24" fill="currentColor">
  <title>LinkedIn icon</title>
  <path d="M20.447 20.452h-3.554v-5.569c0-1.328-.027-3.037-1.852-3.037-1.853 0-2.136 1.445-2.136 2.939v5.667H9.351V9h3.414v1.561h.046c.477-.9 1.637-1.85 3.37-1.85 3.601 0 4.267 2.37 4.267 5.455v6.286zM5.337 7.433c-1.144 0-2.063-.926-2.063-2.065 0-1.138.92-2.063 2.063-2.063 1.14 0 2.064.925 2.064 2.063 0 1.139-.925 2.065-2.064 2.065zm1.782 13.019H3.555V9h3.564v11.452zM22.225 0H1.771C.792 0 0 .774 0 1.729v20.542C0 23.227.792 24 1.771 24h20.451C23.2 24 24 23.227 24 22.271V1.729C24 .774 23.2 0 22.222 0h.003z"/>
</svg>                </div>
              </a>
            </div>
            <div class="w-10-ns w2 ml2 grow">
              <a href="https://twitter.com/chuck_pr" class="light-purple dim link">
                <div class="o-50 pa1-l">
<svg viewBox="0 0 24 24" fill="currentColor">
  <title>Twitter icon</title>
  <path d="M23.954 4.569c-.885.389-1.83.654-2.825.775 1.014-.611 1.794-1.574 2.163-2.723-.951.555-2.005.959-3.127 1.184-.896-.959-2.173-1.559-3.591-1.559-2.717 0-4.92 2.203-4.92 4.917 0 .39.045.765.127 1.124C7.691 8.094 4.066 6.13 1.64 3.161c-.427.722-.666 1.561-.666 2.475 0 1.71.87 3.213 2.188 4.096-.807-.026-1.566-.248-2.228-.616v.061c0 2.385 1.693 4.374 3.946 4.827-.413.111-.849.171-1.296.171-.314 0-.615-.03-.916-.086.631 1.953 2.445 3.377 4.604 3.417-1.68 1.319-3.809 2.105-6.102 2.105-.39 0-.779-.023-1.17-.067 2.189 1.394 4.768 2.209 7.557 2.209 9.054 0 13.999-7.496 13.999-13.986 0-.209 0-.42-.015-.63.961-.689 1.8-1.56 2.46-2.548l-.047-.02z"/>
</svg>                </div>
              </a>
            </div>
    </div>
  </footer>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script>

  console.log('Adding toggle buttons');
  var $toggleCodeButton = $('\
    <a class="toggle toggle-code" href="#0">\
      toggle code\
    </a>\
  ')

  var $toggleAllButton = $('\
    <div>\
      <a class="toggle toggle-all" href="#0">\
        show me\
      </a>\
    </div>\
  ')

  // This adds the buttons
  $(document).ready(function(){

    $cellToggleCode = $('.celltag_toggle_code')
    $cellToggleCode.find('.input').css('display', 'none');
    $toggleCodeButton.prependTo($cellToggleCode);

    $cellToggleAll = $('.celltag_toggle_all');
    $cellToggleAll.css('display', 'none');
    $toggleAllButton.insertBefore($cellToggleAll);

    $('.celltag_disappear').css('display', 'none');

  });

  // These add the toggle behavior for the buttons
  $(document).ready(function(){
    $('a.toggle-code').click(function(){
      $(this).next('.input').slideToggle();
    });
  })

  $(document).ready(function(){
    $('a.toggle-all').click(function(){
      $(this).parent().next('div').slideToggle();
    });
  })

</script></body>

</html>