<!DOCTYPE html>
<html lang="en">

<head>

  <title>Making trees in the Jupyter notebook (part&nbsp;1&nbsp;-&nbsp;IToL)</title>
  <meta itemprop="description"
        name="description"
        content="Let's extend the previous post by scraping the CAZy website for accession numbers from genes …"/>

  
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>

  <!-- Open Graph tags -->
  <meta property="og:locale" content="en_US">
  <meta property="og:site_name" content="A microbiologist with a data science problem"/>
  <meta property="article:author" content="Chuck Pepe-Ranney"/>

  <meta name="twitter:creator" content="@chuckpr"/>

  <link rel="stylesheet" href="https://unpkg.com/tachyons@4.10.0/css/tachyons.min.css"/>

  <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/theme/css/site.css">

  <link rel="shortcut icon" type="image/x-icon" href="https://chuckpr.github.io/favicon.ico"/>
  <link rel="icon" type="image/x-icon" href="https://chuckpr.github.io/favicon.ico"/>



  <!-- Open Graph tags -->
  <meta property="og:type" content="article">
  <meta property="og:title" content="Making trees in the Jupyter notebook (part&nbsp;1&nbsp;-&nbsp;IToL)"/>
  <meta property="og:url" content="https://chuckpr.github.io/posts/trees1/"/>
  <meta name="twitter:card" content="summary_large_image"/>
  <meta property="og:image" content="https://res.cloudinary.com/chuckpr-github-io/image/upload/co_rgb:f2f2f2,w_1000,c_fit,g_east,x_100,l_text:Helvetica_94_right:Making%20trees%20in%20the%20Jupyter%20notebook%20%28part%201%20-%20IToL%29/blog_card.png"/>
  <meta property="article:published_time" content="2015-08-07 00:00:00-04:00"/>
  <meta property="og:description"
        content="Let's extend the previous post by scraping the CAZy website for accession numbers from genes …"/>

  <!-- Twitter card tags -->
  <meta name="twitter:title" content="Making trees in the Jupyter notebook (part&nbsp;1&nbsp;-&nbsp;IToL)">
  <meta name="twitter:description"
        content="Let's extend the previous post by scraping the CAZy website for accession numbers from genes …"/>

  <link rel="canonical" href="https://chuckpr.github.io/posts/trees1/">

  <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/theme/css/pygments.css">
  <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/theme/css/article.css">
    <link rel="stylesheet" type="text/css" href="https://chuckpr.github.io/extra/css/notebook.css">

</head>

<body>
  <header class="flex flex-column avenir mw7 ml-auto mr-auto">
<div class="pt5 pb2 pl0-l pl2">
  <a href="https://chuckpr.github.io/" class="link dim f2-ns f3 fw4 blue tracked-tight lh-title">
    A microbiologist with a data science problem
  </a>
</div>  </header>
  <main class="mw7 ml-auto mr-auto ph3 flex flex-column min-vh-100">
  <article class="flex flex-column">
    <section id="article-title" class="flex justify-between mt4 avenir bb b--moon-gray pb2">
      <div>
        <div class="flex flex-column">
          <div>
            <p class="f3-ns f4 mv0 lh-title">
              Making trees in the Jupyter notebook (part&nbsp;1&nbsp;-&nbsp;IToL)
            </p>
          </div>
          <div class="ml3">
            <p class="f6 ttc mb0 mt1 mid-gray">
              Fri 07 August 2015
            </p>
          </div>
        </div>
      </div>
      <div class="w-40">
        <div class="flex flex-wrap justify-end">
<div class="mv1 mr2 bg-light-green br3 pa2 grow">
  <a href="https://chuckpr.github.io/tag/python" class="link dark-pink underline-hover fw5">
      #Python
  </a>
</div><div class="mv1 mr2 bg-light-green br3 pa2 grow">
  <a href="https://chuckpr.github.io/tag/r" class="link dark-pink underline-hover fw5">
      #R
  </a>
</div><div class="mv1 mr2 bg-light-green br3 pa2 grow">
  <a href="https://chuckpr.github.io/tag/genomics" class="link dark-pink underline-hover fw5">
      #genomics
  </a>
</div>        </div>
      </div>
    </section>
    <section id="article-content" class="ph3 garamond near-black" style="font-size: 17px">
      <p>Let's extend the previous post by scraping the CAZy website for accession numbers from genes in the glycoside hydrolase 5 (GH5) family. Then we can build a phylogenetic tree from these sequences and visualize how well the EC activities correspond to the phylogeny.</p>



<p>Unfortunately the CAZy website spreads the information we need across multiple pages in tables that are not easily parsed. Luckily we have <code>rvest</code>. Let's start by loading the libraries we'll use.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%</span><span class="n">load_ext</span> <span class="n">rpy2</span><span class="o">.</span><span class="n">ipython</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="nf">library</span><span class="p">(</span><span class="n">rvest</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">magrittr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">);</span> <span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">rentrez</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">rwantshue</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>CAZy distinguishes between genes that have been biochemically characterized and those that haven't. For this analysis we'll stick the characterized genes. This is the website that we'll be scraping:<br>
<a href="http://www.cazy.org/GH5_characterized.html">www.cazy.org/GH5_characterized.html</a></p>



<p>All we really need is a table of accessions and corresponding EC numbers but we might as well grab some other information provided by CAZy such as the subfamily membership of each gene and the organism.</p>



<p>We'll start by creating a session from the webpage above.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">cazy</span> <span class="o">=</span> <span class="nf">html_session</span><span class="p">(</span><span class="s">&quot;http://www.cazy.org/GH5_characterized.html&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>We need to grab the links to the pages that contain continued sections for our table. I use Chrome's inspector tools or the <a href="http://selectorgadget.com/">SelectorGadget</a> to find the attributes and element data associated with the information I want to scrape. In this case, the webpage links we want are in span elements of the "pages" class. Each span element has a hyperlink tag ("a" below in <code>html_nodes</code>) and from that tag we want the value of the "href" attribute.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">links</span> <span class="o">=</span> <span class="n">cazy</span> <span class="o">%&gt;%</span> <span class="nf">html_nodes</span><span class="p">(</span><span class="s">&quot;span.pages a&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_attr</span><span class="p">(</span><span class="s">&quot;href&quot;</span><span class="p">)</span>
<span class="n">links</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_text output_subarea ">
<pre>[1] &#34;GH5_characterized.html?debut_PRINC=100#pagination_PRINC&#34;
[2] &#34;GH5_characterized.html?debut_PRINC=200#pagination_PRINC&#34;
[3] &#34;GH5_characterized.html?debut_PRINC=300#pagination_PRINC&#34;
[4] &#34;GH5_characterized.html?debut_PRINC=400#pagination_PRINC&#34;
[5] &#34;GH5_characterized.html?debut_PRINC=500#pagination_PRINC&#34;
</pre>
</div>

</div>

</div>
</div>

</div>

<p>Let's start by scraping the first page (the page we're currently on).</p>



<p>The table has some rows that we don't want (like the row that just points out the Domain of the organisms) which makes parsing this info a bit tricky. We need to scrub rows with the "royaume" class and rows with id "line_titre" (I think those are French words). So, to briefly summarize the command below, we're grabbing rows ("tr") from the table of class "listing" excluding rows of class "royaume" and id "line_titre".</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">rows</span> <span class="o">=</span> <span class="n">cazy</span> <span class="o">%&gt;%</span> <span class="nf">html_nodes</span><span class="p">(</span><span class="s">&quot;table.listing tr:not(.royaume)[id!=line_titre]&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>We'll parse each row to get the accession, EC number, organism, and subfamily. I'm only grabbing the linked accession number. Some of the genes have multiple associated accession numbers but I think we'll only need one. Also, I'm creating one row for each EC number so some of the genes will have two entries. I think I'll eventually remove the genes with multiple EC activities from the tree but it might also be nice to point them out. Note how we use the <code>extract2</code> function from <code>magrittr</code> to get the correct column.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">get_data</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">row</span> <span class="o">%&gt;%</span> <span class="nf">html_nodes</span><span class="p">(</span><span class="s">&quot;td&quot;</span><span class="p">)</span>
    <span class="n">ec</span> <span class="o">=</span> <span class="n">cols</span> <span class="o">%&gt;%</span> <span class="nf">extract2</span><span class="p">(</span><span class="m">2</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_nodes</span><span class="p">(</span><span class="s">&quot;font.E&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_text</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="n">unlist</span>
    <span class="n">organism</span> <span class="o">=</span> <span class="n">cols</span> <span class="o">%&gt;%</span> <span class="nf">extract2</span><span class="p">(</span><span class="m">3</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_text</span><span class="p">(</span><span class="n">trim</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
    <span class="n">acc</span> <span class="o">=</span> <span class="n">cols</span> <span class="o">%&gt;%</span> <span class="nf">extract2</span><span class="p">(</span><span class="m">4</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_nodes</span><span class="p">(</span><span class="s">&quot;a b&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_text</span><span class="p">()</span>
    <span class="nf">if </span><span class="p">(</span> <span class="nf">length</span><span class="p">(</span><span class="n">acc</span><span class="p">)</span> <span class="o">==</span> <span class="m">0</span> <span class="p">)</span>  <span class="p">{</span>
        <span class="n">acc</span> <span class="o">=</span> <span class="n">cols</span> <span class="o">%&gt;%</span> <span class="nf">extract2</span><span class="p">(</span><span class="m">4</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_text</span><span class="p">()</span> <span class="o">%&gt;%</span> <span class="nf">substr</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">10</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">subf</span> <span class="o">=</span> <span class="n">cols</span> <span class="o">%&gt;%</span> <span class="nf">extract2</span><span class="p">(</span><span class="m">7</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">html_text</span><span class="p">(</span><span class="n">trim</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
    <span class="nf">data.frame</span><span class="p">(</span><span class="n">ec</span><span class="o">=</span><span class="n">ec</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span> <span class="n">acc</span><span class="o">=</span><span class="n">acc</span><span class="p">,</span> <span class="n">subf</span><span class="o">=</span><span class="n">subf</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>And now we'll <code>ldply</code> it into a handy dataframe.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">df.pg1</span> <span class="o">=</span> <span class="nf">ldply</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">get_data</span><span class="p">)</span>
<span class="n">df.pg1</span> <span class="o">%&gt;%</span> <span class="n">head</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_text output_subarea ">
<pre>        ec                                   organism        acc subf
1  3.2.1.4        Desulfurococcaceae archaeon EBI-244 AEB53062.1     
2  3.2.1.-            Halorhabdus utahensis DSM 12940 ACV12553.1     
3  3.2.1.4                      Pyrococcus abyssi GE5 CAB49854.1    1
4  3.2.1.4                  Pyrococcus horikoshii OT3 AAQ31833.1    1
5 3.2.1.78                Acidothermus cellulolyticus ABJ41262.1    7
6  3.2.1.4 Acidothermus cellulolyticus 11B ATCC 43068 AAA75477.1    1
</pre>
</div>

</div>

</div>
</div>

</div>

<p>Now we need to do that for all the other pages of the table. This function uses <code>jump_to</code> to go to a link and extract the information we need with the same code as we had above. We can append our page 1 information with <code>rbind</code> and we're ready to go.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">get_df</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">link</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">session</span> <span class="o">=</span> <span class="nf">jump_to</span><span class="p">(</span><span class="n">cazy</span><span class="p">,</span> <span class="n">link</span><span class="p">)</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="n">session</span> <span class="o">%&gt;%</span> <span class="nf">html_nodes</span><span class="p">(</span><span class="s">&quot;table.listing tr:not(.royaume)[id!=line_titre]&quot;</span><span class="p">)</span>
    <span class="n">rows</span> <span class="o">%&gt;%</span> <span class="n">head</span>
    <span class="nf">ldply</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">get_data</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">df.all</span> <span class="o">=</span> <span class="nf">ldply</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">get_df</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">rbind</span><span class="p">(</span><span class="n">df.pg1</span><span class="p">)</span>
<span class="n">df.all</span> <span class="o">%&gt;%</span> <span class="n">head</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_text output_subarea ">
<pre>        ec                             organism        acc subf
1 3.2.1.78       Clostridium cellulovorans 743B ADL52789.1    7
2 3.2.1.78       Clostridium cellulovorans 743B AAF06110.2   17
3  3.2.1.4 Clostridium cellulovorans ATCC 35269 AAA23233.1    4
4 3.2.1.73 Clostridium cellulovorans ATCC 35269 AAA23233.1    4
5  3.2.1.8 Clostridium cellulovorans ATCC 35269 AAA23233.1    4
6  3.2.1.4 Clostridium cellulovorans ATCC 35269 AAD39739.1    4
</pre>
</div>

</div>

</div>
</div>

</div>

<p>We now have all the information we need to make our tree in IToL that will give us a sense of how phylogenetically conserved the EC annotations are with phylogeny. But first let's plot some counts of the different EC activities. There is clearly one predominant activity for this set of sequences.</p>


<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">h</span> <span class="m">300</span> <span class="o">-</span><span class="n">w</span> <span class="m">600</span>
<span class="n">df.all</span> <span class="o">%&gt;%</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">ec</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">summarize</span><span class="p">(</span><span class="n">count</span> <span class="o">=</span> <span class="nf">n</span><span class="p">())</span> <span class="o">%&gt;%</span> <span class="p">{</span>
        <span class="n">xsort</span> <span class="o">=</span> <span class="nf">arrange</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="nf">desc</span><span class="p">(</span><span class="n">count</span><span class="p">))</span> <span class="o">%&gt;%</span> <span class="nf">extract2</span><span class="p">(</span><span class="s">&quot;ec&quot;</span><span class="p">)</span>
        <span class="nf">mutate</span><span class="p">(</span><span class="n">.</span><span class="p">,</span> <span class="n">ec</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">ec</span><span class="p">,</span> <span class="n">levels</span> <span class="o">=</span> <span class="n">xsort</span><span class="p">))</span> <span class="o">%&gt;%</span>
        <span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">ec</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">count</span><span class="p">))</span> <span class="o">+</span>
            <span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span> <span class="o">=</span> <span class="s">&quot;identity&quot;</span><span class="p">)</span> <span class="o">+</span> 
            <span class="nf">theme_bw</span><span class="p">()</span> <span class="o">+</span>
            <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span> <span class="o">=</span> <span class="m">45</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">),</span>
                  <span class="n">axis.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">16</span><span class="p">))</span>
    <span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/trees1_20_0.png"
>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">df.all</span> <span class="o">%&gt;%</span> <span class="nf">write.csv</span><span class="p">(</span><span class="s">&quot;data/GH5_data.csv&quot;</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>We should add one more bit of information to the dataframe, the colors for each EC number. We need to send this information to IToL. Below I'm using <a href="https://github.com/hoesler/rwantshue">rwantshue</a> to get a list of colors the same length as the number of observed EC numbers. We collect the colors for each EC number in a dataframe and then we can join that colors dataframe to our CAZy dataframe. We'll also save the IToL information to a file. I am making what IToL refers to as a "colors definition" file. This file should include (in order) a column for the tip names on the tree (in our case this will be accessions), a column specifying what you want to color ("range" here see IToL website for details), the actual color, and the label (EC numbers in our case).</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">N</span> <span class="o">=</span> <span class="n">df.all</span><span class="o">$</span><span class="n">ec</span> <span class="o">%&gt;%</span> <span class="n">unique</span> <span class="o">%&gt;%</span> <span class="n">length</span>
<span class="n">scheme</span> <span class="o">&lt;-</span> <span class="nf">iwanthue</span><span class="p">()</span>
<span class="n">colors</span> <span class="o">=</span> <span class="n">scheme</span><span class="o">$</span><span class="nf">hex</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">color_space</span> <span class="o">=</span> <span class="n">hcl_presets</span><span class="o">$</span><span class="n">intense</span><span class="p">)</span>
<span class="n">col.df</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">colors</span><span class="p">,</span>
    <span class="n">ec</span> <span class="o">=</span> <span class="n">df.all</span><span class="o">$</span><span class="n">ec</span> <span class="o">%&gt;%</span> <span class="n">unique</span>
<span class="p">)</span>
<span class="n">df.all</span> <span class="o">%&gt;%</span> 
    <span class="nf">left_join</span><span class="p">(</span><span class="n">col.df</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">ctype</span> <span class="o">=</span> <span class="s">&quot;range&quot;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">ec</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">select</span><span class="p">(</span><span class="n">acc</span><span class="p">,</span> <span class="n">ctype</span><span class="p">,</span> <span class="n">colors</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">acc</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">filter</span><span class="p">(</span><span class="nf">row_number</span><span class="p">()</span> <span class="o">==</span> <span class="m">1</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="c1"># just grab the first row for each seq. </span>
                                                    <span class="c1"># Some seqs have two EC numbers.</span>
    <span class="nf">write.table</span><span class="p">(</span><span class="s">&quot;data/color_defs.tsv&quot;</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">col.names</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot;\t&quot;</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Here's what that file looks like,</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="err">!</span><span class="n">head</span> <span class="o">-</span><span class="n">n5</span> <span class="n">data</span><span class="o">/</span><span class="n">color_defs</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>ADL52789.1	range	#4FB197	3.2.1.78
AAF06110.2	range	#4FB197	3.2.1.78
AAA23233.1	range	#CA55E1	3.2.1.4
AAD39739.1	range	#CA55E1	3.2.1.4
BAA12826.1	range	#CA55E1	3.2.1.4
</pre>
</div>
</div>

</div>
</div>

</div>

<p>Great! Now on to building the tree...</p>



<p>We need to get the sequences first. Let's use <a href="https://github.com/ropensci/rentrez">rentrez</a> to query Genbank for our accessions.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<span class="n">accs</span> <span class="o">=</span> <span class="n">df.all</span><span class="o">$</span><span class="n">acc</span> <span class="o">%&gt;%</span> <span class="n">as.character</span> <span class="o">%&gt;%</span> <span class="n">unique</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="nf">entrez_fetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">id</span><span class="o">=</span><span class="n">accs</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nf">write</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="s">&quot;data/GH5_characterized.fasta&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="err">!</span><span class="n">head</span> <span class="o">-</span><span class="n">n3</span> <span class="n">data</span><span class="o">/</span><span class="n">GH5_characterized</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>&gt;gi|328834886|gb|AEB53062.1| putative glycoside hydrolase family 5 [Desulfurococcaceae archaeon EBI-244]
MKKVHIIAIVVIIAIAFALILARYYTMQRGYETVTPTTPPQQTTTTETTPVPTEAGTTTPITEATVTQPP
QTPTTPSPQTPTTPTALPTPSPTPTAPSATVTETTSPQTPTTTITTETTTTPAPQPQVVFLKLPEGEEPK
</pre>
</div>
</div>

</div>
</div>

</div>

<p>I'm going to align these sequences with <a href="http://www.drive5.com/muscle/">Muscle</a>.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="err">!</span><span class="n">muscle</span> <span class="o">-</span><span class="ow">in</span> <span class="n">data</span><span class="o">/</span><span class="n">GH5_characterized</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">data</span><span class="o">/</span><span class="n">GH5_characterized</span><span class="o">.</span><span class="n">aln</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="err">!</span><span class="n">head</span> <span class="o">-</span><span class="n">n5</span> <span class="n">data</span><span class="o">/</span><span class="n">GH5_characterized</span><span class="o">.</span><span class="n">aln</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>&gt;gi|190684977|gb|ACE82655.1| endo-1, 4-beta mannanase, man5C [Cellvibrio japonicus Ueda107]
------------------------------------------------------------
------------------------------------------------------------
------------------------------------------------------------
-------------------------------------------MKHLISPRILTGLFVIG
</pre>
</div>
</div>

</div>
</div>

</div>

<p>Ok, we're going to switch into Python mode now as we have a look at this alignment. Also, we'll use Python to talk to IToL. Let's just plot gap frequency by position.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="kn">from</span> <span class="nn">skbio.sequence</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">itolapi</span> <span class="kn">import</span> <span class="n">Itol</span><span class="p">,</span> <span class="n">ItolExport</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s2">&quot;data/GH5_characterized.aln.fasta&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Our alignment is a bit gappy as you can see below. There are quite a few positions with few characters.</p>


<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gap_freq</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span><span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">position_frequencies</span><span class="p">()]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">([</span><span class="mi">16</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gap_freq</span><span class="p">)),</span> <span class="n">gap_freq</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Position&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Gap frequency&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt output_prompt">Out[145]:</div>




<div class="output_text output_subarea output_execute_result">
<pre>&lt;matplotlib.text.Text at 0x7f073bdc0e50&gt;</pre>
</div>

</div>

<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/trees1_37_1.png"
>
</div>

</div>

</div>
</div>

</div>

<p>Let's make a gap frequency filter to remove columns with greater than 60% gaps.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">keepers</span><span class="p">,</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gap_freq</span><span class="p">)</span><span class="o">&lt;=</span><span class="mf">0.60</span><span class="p">)</span>
<span class="n">aln_masked</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">subalignment</span><span class="p">(</span><span class="n">positions_to_keep</span><span class="o">=</span><span class="n">keepers</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>

<p>Ok, that looks better. We'll use this filtered alignment to build the tree.</p>


<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gap_freq2</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span><span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">aln_masked</span><span class="o">.</span><span class="n">position_frequencies</span><span class="p">()]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">([</span><span class="mi">16</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gap_freq2</span><span class="p">)),</span> <span class="n">gap_freq2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Position&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Gap frequency&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt output_prompt">Out[147]:</div>




<div class="output_text output_subarea output_execute_result">
<pre>&lt;matplotlib.text.Text at 0x7f073bf9dd10&gt;</pre>
</div>

</div>

<div class="output_area">

    <div class="prompt"></div>




<div class="output_png output_subarea ">
<img src="https://chuckpr.github.io/notebooks/images/trees1_41_1.png"
>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">aln_masked</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;data/GH5_characterized.aln.masked.fasta&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt output_prompt">Out[148]:</div>




<div class="output_text output_subarea output_execute_result">
<pre>&#39;data/GH5_characterized.aln.masked.fasta&#39;</pre>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="err">!</span><span class="n">head</span> <span class="o">-</span><span class="n">n2</span> <span class="n">data</span><span class="o">/</span><span class="n">GH5_characterized</span><span class="o">.</span><span class="n">aln</span><span class="o">.</span><span class="n">masked</span><span class="o">.</span><span class="n">fasta</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">c1</span><span class="o">-</span><span class="mi">100</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>&gt;gi|190684977|gb|ACE82655.1| endo-1, 4-beta mannanase, man5C [Cellvibrio japonicus Ueda107]
HLISPRILTGLFVIGLAPAWSSSVSSSGIIVGAQCNWYGPVCQSGASLPAPGNSVL-SGSSPEGNSWLSPYS-SKVMDGTASYAFAEATVPVTYQGQGDG
</pre>
</div>
</div>

</div>
</div>

</div>

<p>I'm going to make the tree with <a href="http://www.microbesonline.org/fasttree/">FastTree</a>. I'm doing a quick hack to make sure that the tips of our tree are just accession numbers and not the long identifiers returned from <code>rentrez</code>. The sed command just grabs the accession from each FASTA identifier line and scrubs the rest. I am using <a href="http://tldp.org/LDP/abs/html/process-sub.html">process substitution</a> but you could just as easily write the scrubbed FASTA to a file and send it to FastTree.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="err">!</span><span class="n">head</span> <span class="o">-</span><span class="n">n6</span> <span class="n">data</span><span class="o">/</span><span class="n">GH5_characterized</span><span class="o">.</span><span class="n">aln</span><span class="o">.</span><span class="n">masked</span><span class="o">.</span><span class="n">fasta</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;/^&gt;/ s/.*|.*|.*|\(.*\)|.*/&gt;</span><span class="se">\1</span><span class="s1">/&#39;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">c</span> <span class="mi">1</span><span class="o">-</span><span class="mi">80</span> <span class="c1">#for example...</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>&gt;ACE82655.1
HLISPRILTGLFVIGLAPAWSSSVSSSGIIVGAQCNWYGPVCQSGASLPAPGNSVL-SGSSPEGNSWLSPYS-SKVMDGT
&gt;AFR99035.1
----MPPPPEVSPVTG-------------NPVSPHYIHS------STL-HFGRSLVLRGVN--EGFWFPLNLDDGSLARL
&gt;BAG70961.1
MKAMTSVILPPVAAPALCLAAV---------WTTFNAVG------ANL-G-GW---LIDTTGAVDEWCGPVLEWITIDTL
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span>%%bash
FastTree &lt;<span class="o">(</span>sed <span class="s1">&#39;/^&gt;/ s/.*|.*|.*|\(.*\)|.*/&gt;\1/&#39;</span> data/GH5_characterized.aln.masked.fasta<span class="o">)</span> &gt; data/GH5.tree <span class="m">2</span>&gt;/dev/null
</pre></div>

    </div>
</div>
</div>

</div>

<p>Cool, now we just need to upload the tree and the color definition file from above to IToL. There is a nice Python library to talk to IToL programatically called <a href="https://github.com/albertyw/itolapi">itolapi</a>. We're using that to upload our info. The <code>itolapi</code> page has all the documentation required to use it effectively.</p>


<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">itol_uploader</span> <span class="o">=</span> <span class="n">Itol</span><span class="o">.</span><span class="n">Itol</span><span class="p">()</span>
<span class="n">itol_uploader</span><span class="o">.</span><span class="n">add_variable</span><span class="p">(</span><span class="s1">&#39;treeFile&#39;</span><span class="p">,</span> <span class="s1">&#39;data/GH5.tree&#39;</span><span class="p">)</span>
<span class="n">itol_uploader</span><span class="o">.</span><span class="n">add_variable</span><span class="p">(</span><span class="s1">&#39;treeFormat&#39;</span><span class="p">,</span> <span class="s1">&#39;newick&#39;</span><span class="p">)</span>
<span class="n">itol_uploader</span><span class="o">.</span><span class="n">add_variable</span><span class="p">(</span><span class="s1">&#39;treeName&#39;</span><span class="p">,</span> <span class="s1">&#39;GH5&#39;</span><span class="p">)</span>
<span class="n">itol_uploader</span><span class="o">.</span><span class="n">add_variable</span><span class="p">(</span><span class="s2">&quot;colorDefinitionFile&quot;</span><span class="p">,</span> <span class="s2">&quot;data/color_defs.tsv&quot;</span><span class="p">)</span>
<span class="n">status</span> <span class="o">=</span> <span class="n">itol_uploader</span><span class="o">.</span><span class="n">upload</span><span class="p">()</span>
<span class="n">itol_uploader</span><span class="o">.</span><span class="n">comm</span><span class="o">.</span><span class="n">upload_output</span>
<span class="nb">print</span> <span class="n">itol_uploader</span><span class="o">.</span><span class="n">get_webpage</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>http://itol.embl.de/external.cgi?tree=128841252312946714389719820&amp;restore_saved=1
</pre>
</div>
</div>

</div>
</div>

</div>

<p>You can also use <code>itolapi</code> to download the tree but I don't think the IToL API supports exporting trees with full clade coloring. I could be wrong but I really wanted the full clade coloring as opposed to just a ring of colors around the tree. Anyway, I had to manually visit IToL to export this visualization (frown).</p>


<div class="cell border-box-sizing code_cell rendered celltag_fullwidth">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">Image</span><span class="p">(</span><span class="s2">&quot;data/GH5.tree.png&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt output_prompt">Out[16]:</div>




<div class="output_png output_subarea output_execute_result">
<img src="https://chuckpr.github.io/notebooks/images/trees1_50_0.png"
width=600
>
</div>

</div>

</div>
</div>

</div>

<p>So it looks like there's some coherence to the EC numbers. The two main EC annotations do not appear to overlap much but the minor annotations are dispersed throughout in small groups.</p>



<p>You can explore this tree in IToL by visiting <a href="http://itol.embl.de/shared/chuckpr">this page</a>.</p>



<p>Disclaimer: I don't want to go down the rabbit hole of finding all the inconsistencies in this CAZy table for this small example. There are most definitely some wonky rows in the information we scraped. I think the accessions are the least consistent field in CAZy and it's tiresome to account for all the seemingly arbitrary little ways that rows can differ from the consensus. The moral of the story? <strong>Don't make people scrape your webpage to get information!!!</strong> I don't see why that table can't be provided by the CAZy team as a nice CSV...</p>


 


    </section>
<div id="disqus_thread" class="mt4">
</div>
<script>

  var disqus_config = function () {
    this.page.url = 'https://chuckpr.github.io/posts/trees1/';
    this.page.identifier = 'posts/trees1/';
  };

  (function() { // DON'T EDIT BELOW THIS LINE
    var d = document, s = d.createElement('script');
    s.src = 'https://chuckpr-blog.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();

</script>
<noscript>
  Please enable JavaScript to view the
  <a href="https://disqus.com/?ref_noscript">
    comments powered by Disqus.
  </a>
</noscript>  </article>
  </main>
  <footer class="pv3 flex justify-between mt4 mb2 mw7 ml-auto mr-auto avenir">
    <div class="flex flex-column pl3">
      <div class="mid-gray fw4 f6-ns f7 mb2">
        Copyright &copy; 2020 Chuck&nbsp;Pepe&#8209;Ranney
      </div>
      <div class="mid-gray" style="font-size: 0.65rem;">
        Powered&nbsp;by&nbsp;<a href="https://getpelican.com/" class="link dim
          mid-gray f6 fw6">Pelican</a>,
        Theme&nbsp;by&nbsp;<a href="https://github.com/chuckpr/simple-pelican-blog-theme"
          class="link dim mid-gray fw6">@chuckpr</a>
      </div>
    </div>
    <div class="flex justify-end w-50-ns w-30 mr4-ns mr0 ml2">
            <div class="w-10-ns w2 ml2 grow">
              <a href="https://github.com/chuckpr" class="light-purple dim link">
                <div class="o-50 pa1-l">
<svg viewBox="0 0 24 24" fill="currentColor">
  <title>GitHub icon</title>
  <path d="M12 .297c-6.63 0-12 5.373-12 12 0 5.303 3.438 9.8 8.205 11.385.6.113.82-.258.82-.577 0-.285-.01-1.04-.015-2.04-3.338.724-4.042-1.61-4.042-1.61C4.422 18.07 3.633 17.7 3.633 17.7c-1.087-.744.084-.729.084-.729 1.205.084 1.838 1.236 1.838 1.236 1.07 1.835 2.809 1.305 3.495.998.108-.776.417-1.305.76-1.605-2.665-.3-5.466-1.332-5.466-5.93 0-1.31.465-2.38 1.235-3.22-.135-.303-.54-1.523.105-3.176 0 0 1.005-.322 3.3 1.23.96-.267 1.98-.399 3-.405 1.02.006 2.04.138 3 .405 2.28-1.552 3.285-1.23 3.285-1.23.645 1.653.24 2.873.12 3.176.765.84 1.23 1.91 1.23 3.22 0 4.61-2.805 5.625-5.475 5.92.42.36.81 1.096.81 2.22 0 1.606-.015 2.896-.015 3.286 0 .315.21.69.825.57C20.565 22.092 24 17.592 24 12.297c0-6.627-5.373-12-12-12"/>
</svg>                </div>
              </a>
            </div>
            <div class="w-10-ns w2 ml2 grow">
              <a href="https://www.linkedin.com/in/chuckpr/" class="light-purple dim link">
                <div class="o-50 pa1-l">
<svg viewBox="0 0 24 24" fill="currentColor">
  <title>LinkedIn icon</title>
  <path d="M20.447 20.452h-3.554v-5.569c0-1.328-.027-3.037-1.852-3.037-1.853 0-2.136 1.445-2.136 2.939v5.667H9.351V9h3.414v1.561h.046c.477-.9 1.637-1.85 3.37-1.85 3.601 0 4.267 2.37 4.267 5.455v6.286zM5.337 7.433c-1.144 0-2.063-.926-2.063-2.065 0-1.138.92-2.063 2.063-2.063 1.14 0 2.064.925 2.064 2.063 0 1.139-.925 2.065-2.064 2.065zm1.782 13.019H3.555V9h3.564v11.452zM22.225 0H1.771C.792 0 0 .774 0 1.729v20.542C0 23.227.792 24 1.771 24h20.451C23.2 24 24 23.227 24 22.271V1.729C24 .774 23.2 0 22.222 0h.003z"/>
</svg>                </div>
              </a>
            </div>
            <div class="w-10-ns w2 ml2 grow">
              <a href="https://twitter.com/chuck_pr" class="light-purple dim link">
                <div class="o-50 pa1-l">
<svg viewBox="0 0 24 24" fill="currentColor">
  <title>Twitter icon</title>
  <path d="M23.954 4.569c-.885.389-1.83.654-2.825.775 1.014-.611 1.794-1.574 2.163-2.723-.951.555-2.005.959-3.127 1.184-.896-.959-2.173-1.559-3.591-1.559-2.717 0-4.92 2.203-4.92 4.917 0 .39.045.765.127 1.124C7.691 8.094 4.066 6.13 1.64 3.161c-.427.722-.666 1.561-.666 2.475 0 1.71.87 3.213 2.188 4.096-.807-.026-1.566-.248-2.228-.616v.061c0 2.385 1.693 4.374 3.946 4.827-.413.111-.849.171-1.296.171-.314 0-.615-.03-.916-.086.631 1.953 2.445 3.377 4.604 3.417-1.68 1.319-3.809 2.105-6.102 2.105-.39 0-.779-.023-1.17-.067 2.189 1.394 4.768 2.209 7.557 2.209 9.054 0 13.999-7.496 13.999-13.986 0-.209 0-.42-.015-.63.961-.689 1.8-1.56 2.46-2.548l-.047-.02z"/>
</svg>                </div>
              </a>
            </div>
    </div>
  </footer>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script>

  console.log('Adding toggle buttons');
  var $toggleCodeButton = $('\
    <a class="toggle toggle-code" href="#0">\
      toggle code\
    </a>\
  ')

  var $toggleAllButton = $('\
    <div>\
      <a class="toggle toggle-all" href="#0">\
        show me\
      </a>\
    </div>\
  ')

  // This adds the buttons
  $(document).ready(function(){

    $cellToggleCode = $('.celltag_toggle_code')
    $cellToggleCode.find('.input').css('display', 'none');
    $toggleCodeButton.prependTo($cellToggleCode);

    $cellToggleAll = $('.celltag_toggle_all');
    $cellToggleAll.css('display', 'none');
    $toggleAllButton.insertBefore($cellToggleAll);

    $('.celltag_disappear').css('display', 'none');

  });

  // These add the toggle behavior for the buttons
  $(document).ready(function(){
    $('a.toggle-code').click(function(){
      $(this).next('.input').slideToggle();
    });
  })

  $(document).ready(function(){
    $('a.toggle-all').click(function(){
      $(this).parent().next('div').slideToggle();
    });
  })

</script></body>

</html>